core/src/main/scala/org/genivi/sota/core/transfer/VehicleUpdates.scala

```diff
@@ -63,45 +58,91 @@ object VehicleUpdates {
+
+    // look up the namespace
+    val namespace: Namespace = {
+      val namespaceQuery =
+        updateSpecs
+          .filter(r => r.requestId === updateReport.update_id)
+          .map(_.namespace)
+      //TODO: get default namespace for now, this should be replaced with a lookup to the updates table eventually
+      NamespaceDirective.defaultNs.get
+    }
+
+    // add a row (with fresh UUID) to OperationResult table for each result of this report
+    val writeResultsIO: List[DBIO[OperationResult]] = for {
+      rviOpResult <- updateReport.operation_results
+      dbOpResult   = OperationResult.from(
+        rviOpResult,
+        updateReport.update_id,
+        vin,
+        namespace)
+    } yield OperationResults.persist(dbOpResult)
+
+    // add a row (with auto-inc PK) to InstallHistory table
+    def writeHistoryIO(spec: UpdateSpec): DBIO[Int] = {
+      InstallHistories.log(
+        spec.namespace, vin,
+        spec.request.id, spec.request.packageId,
+        success = updateReport.isSuccess)
+    }
+
+    // look up the UpdateSpec to rewrite its status and to use it as FK in InstallHistory
+    val newStatus = if (updateReport.isSuccess) UpdateStatus.Finished else UpdateStatus.Failed
     val dbIO = for {
       spec <- findUpdateSpecFor(vin, updateReport.update_id)
-      _ <- DBIO.sequence(writeResultsIO)
-      _ <- UpdateSpecs.setStatus(spec, UpdateStatus.Finished)
-      _ <- InstallHistories.log(spec.namespace, vin, spec.request.id, spec.request.packageId, success = wasSuccessful)
-    } yield spec.copy(status = UpdateStatus.Finished)
+      _    <- DBIO.sequence(writeResultsIO)
+      _    <- UpdateSpecs.setStatus(spec, newStatus)
+      _    <- writeHistoryIO(spec)
+      _    <- cancelInstallationQueueIO(vin, spec, updateReport.isFail, updateReport.update_id)
+    } yield spec.copy(status = newStatus)
 
     db.run(dbIO.transactionally)
   }
 
+  /**
+    * <ul>
+    *   <li>
+    *     For a failed UpdateReport, mark as cancelled the rest of the installation queue.
+    *     <ul>
+    *       <li>"The rest of the installation queue" defined as those (InFlight and Pending) UpdateSpec-s
+    *       with (strictly) greater installPos than the given one, for the VIN in question.</li>
+    *       <li>Note: those UpdateSpec-s correspond to different UpdateRequests than the current one.</li>
+    *     </ul>
+    *   </li>
+    *   <li>For a successful UpdateReport, do nothing.</li>
+    * </ul>
+    */
+  def cancelInstallationQueueIO(vin: Vehicle.Vin,
+                                spec: UpdateSpec,
+                                isFail: Boolean,
+                                updateRequestId: UUID): DBIO[Int] = {
+    updateSpecs
+      .filter(_.vin === vin && isFail)
+      .filter(_.requestId =!= updateRequestId)
+      .filter(_.status.inSet(List(UpdateStatus.InFlight, UpdateStatus.Pending)))
+      .filter(_.installPos > spec.installPos)
```

Comment:
It's not a bug elsewhere, the installPos exists to define an installation order, but there is a default installation order: FIFO

I don't unsertand how the behaviour is still ok? This will not cancel subsequent updates if they all have installPos == 0